File: src/tests/eina_js/eina_js_suite.cc

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <Eina.hh>
#include EINA_STRINGIZE(V8_INCLUDE_HEADER)

#include <cassert>
#include <cstdlib>
#include <fstream>

#include <Eina.h>
#include <Eo.hh>

#include <eina_js_accessor.hh>
#include <eina_js_list.hh>
#include <eina_js_iterator.hh>
#include <eina_js_error.hh>
#include <eina_js_log.cc>

const char* ToCString(const v8::String::Utf8Value& value) {
  return *value ? *value : "<string conversion failed>";
}

// Executes a string within the current v8 context.
bool ExecuteString(v8::Isolate* isolate,
                   v8::Handle<v8::String> source,
                   v8::Handle<v8::Value> name)
{
  efl::eina::js::compatibility_handle_scope handle_scope(isolate);
  v8::TryCatch try_catch;
  v8::ScriptOrigin origin(name);
  v8::Handle<v8::Script> script = v8::Script::Compile(source, &origin);
  if (script.IsEmpty()) {
    std::cerr << "Compilation failed" << std::endl;
    std::abort();
    // Print errors that happened during compilation.
    // if (report_exceptions)
    //   ReportException(isolate, &try_catch);
    return false;
  }
  else
  {
    std::cerr << "Compilation succesful" << std::endl;
    v8::Handle<v8::Value> result = script->Run();
    if (result.IsEmpty()) {
      std::cout << "Failed with exception thrown" << std::endl;
      //assert(try_catch.HasCaught());
      //std::abort();
      // Print errors that happened during execution.
      // if (report_exceptions)
      //   ReportException(isolate, &try_catch);
      if(try_catch.HasCaught())
        std::cerr << "Exception " << ToCString(v8::String::Utf8Value(try_catch.Message()->Get()))
                  << std::endl;
      std::abort();
      return false;
    } else {
      assert(!try_catch.HasCaught());
      // if (print_result && !result->IsUndefined()) {
      //   // If all went well and the result wasn't undefined then print
      //   // the returned value.
      //   v8::String::Utf8Value str(result);
      //   const char* cstr = ToCString(str);
      //   printf("%s\n", cstr);
      // }
      return true;
    }
  }
}

efl::eina::js::compatibility_return_type Print(efl::eina::js::compatibility_callback_info_type args)
{
  bool first = true;
  for (int i = 0; i < args.Length(); i++) {
    efl::eina::js::compatibility_handle_scope handle_scope(args.GetIsolate());
    if (first) {
      first = false;
    } else {
      printf(" ");
    }
    v8::String::Utf8Value str(args[i]);
    const char* cstr = ToCString(str);
    printf("%s", cstr);
  }
  printf("\n");
  fflush(stdout);
  return efl::eina::js::compatibility_return();
}

efl::eina::js::compatibility_return_type clear_eina_error(efl::eina::js::compatibility_callback_info_type args)
{
    eina_error_set(0);
    return efl::eina::js::convert_error_to_javascript_exception(args.GetIsolate());
}

efl::eina::js::compatibility_return_type set_eina_error(efl::eina::js::compatibility_callback_info_type args)
{
    eina_error_set(eina_error_msg_static_register("foobar"));
    return efl::eina::js::convert_error_to_javascript_exception(args.GetIsolate());
}

EAPI void eina_container_register(v8::Handle<v8::Object> global, v8::Isolate* isolate);
EAPI v8::Local<v8::Function> get_list_instance_template();

efl::eina::ptr_list<int> list;
efl::eina::js::range_eina_list<int> raw_list;

efl::eina::array<int> array;

static void eina_log_print_cb_js_test(const Eina_Log_Domain *d,
                                      Eina_Log_Level level, const char *file,
                                      const char *fnc, int line,
                                      const char */*fmt*/, void */*data*/,
                                      va_list args)
{
    using std::string;

    static int index = 0;

    const char *msg = va_arg(args, const char*);

    string domains[] = {"", "", "", "mydomain", "mydomain2"};
    int levels[] = {EINA_LOG_LEVEL_DBG, EINA_LOG_LEVEL_CRITICAL,
                    EINA_LOG_LEVEL_WARN, EINA_LOG_LEVEL_INFO,
                    EINA_LOG_LEVEL_ERR};
    string functions[] = {"f1", "", "f2", "f3", ""};
#ifdef HAVE_NODEJS
    int lines[] = {191, 192, 193, 194, 200};
#else
    int lines[] = {190, 191, 192, 193, 199};
#endif
    string messages[] = {"I changed again", "Cool to Hate", "One Fine Day",
                         "Never Gonna Find Me", "The Kids Aren't Alright"};

    assert(string(d->name, d->namelen) == domains[index]);
    assert(level == levels[index]);
    {
        auto last_component = strrchr(file, '/');
        assert(last_component);
        assert(last_component == string("/eina_js_suite.js"));
    }
    assert(fnc == functions[index]);
    assert(line == lines[index]);
    assert(msg == messages[index]);
    ++index;
}

void test_setup(v8::Handle<v8::Object> exports)
{
  v8::Isolate* isolate = v8::Isolate::GetCurrent();

  // container globals
  std::cerr << __LINE__ << std::endl;
  list.push_back(new int(5));
  list.push_back(new int(10));
  list.push_back(new int(15));
  std::cerr << __LINE__ << std::endl;

  raw_list = efl::eina::js::range_eina_list<int>(list.native_handle());

  std::cerr << __LINE__ << std::endl;
  eina_container_register(exports, v8::Isolate::GetCurrent());
  std::cerr << __LINE__ << std::endl;
  
  v8::Handle<v8::Value> a[] = {efl::eina::js::compatibility_new<v8::External>(nullptr, &raw_list)};
  std::cerr << __LINE__ << std::endl;
  exports->Set(efl::eina::js::compatibility_new<v8::String>(nullptr, "raw_list")
               , get_list_instance_template()->NewInstance(1, a));
  std::cerr << __LINE__ << std::endl;

  // error globals
  exports->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "clear_eina_error"),
              efl::eina::js::compatibility_new<v8::FunctionTemplate>(isolate, clear_eina_error)
              ->GetFunction());
  exports->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "set_eina_error"),
              efl::eina::js::compatibility_new<v8::FunctionTemplate>(isolate, set_eina_error)
              ->GetFunction());
  std::cerr << __LINE__ << std::endl;

  // accessor
  Eina_Array* array = eina_array_new(3);
  eina_array_push(array, new int(42));
  eina_array_push(array, new int(24));
  eina_array_push(array, new int(0));

  static efl::eina::accessor<int> acc(eina_array_accessor_new(array));
  std::cerr << __LINE__ << std::endl;
  
  v8::Local<v8::Object> wrapped_acc = efl::eina::js::export_accessor( acc, isolate);
  exports->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "acc"), wrapped_acc);

  static efl::eina::iterator<int> it(eina_array_iterator_new(array));

  v8::Local<v8::Object> wrapped_it = efl::eina::js::export_iterator(&it, isolate);

  exports->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "it"), wrapped_it);
  
  std::cerr << __LINE__ << std::endl;

  efl::eina::js::register_value
    (isolate, exports
     , efl::eina::js::compatibility_new<v8::String>(isolate, "value"));
  
  std::cerr << __LINE__ << std::endl;

  // log
  using namespace efl::eina::js;
  using v8::String;
  {
      int d = eina_log_domain_register("mydomain", "");
      eina_log_domain_registered_level_set(d, EINA_LOG_LEVEL_DBG);
      exports->Set(compatibility_new<String>(isolate, "mydomain"),
                   value_cast<v8::Local<v8::Value>>(d, isolate));
  }
  register_log_level_critical(isolate, exports,
                              compatibility_new<String>(isolate,
                                                        "LOG_LEVEL_CRITICAL"));
  register_log_level_err(isolate, exports,
                         compatibility_new<String>(isolate, "LOG_LEVEL_ERR"));
  register_log_level_warn(isolate, exports,
                          compatibility_new<String>(isolate, "LOG_LEVEL_WARN"));
  register_log_level_info(isolate, exports,
                          compatibility_new<String>(isolate, "LOG_LEVEL_INFO"));
  register_log_level_dbg(isolate, exports,
                         compatibility_new<String>(isolate, "LOG_LEVEL_DBG"));
  register_log_domain_global(isolate, exports,
                             compatibility_new<String>(isolate,
                                                       "LOG_DOMAIN_GLOBAL"));
  register_log_print(isolate, exports,
                     compatibility_new<String>(isolate, "log_print"));
  register_log_domain_register(isolate, exports,
                               compatibility_new<String>(isolate,
                                                         "log_domain"
                                                         "_register"));
  register_log_domain_unregister(isolate, exports,
                                 compatibility_new<String>(isolate,
                                                           "log_domain_unregis"
                                                           "ter"));
  register_log_domain_registered_level_get(isolate, exports,
                                           compatibility_new<String>(isolate,
                                                                     "log"
                                                                     "_domain"
                                                                     "_register"
                                                                     "ed_level"
                                                                     "_get"));
  register_log_domain_registered_level_set(isolate, exports,
                                           compatibility_new<String>(isolate,
                                                                     "log"
                                                                     "_domain"
                                                                     "_register"
                                                                     "ed_level"
                                                                     "_set"));
  register_log_print_cb_set(isolate, exports,
                            compatibility_new<String>(isolate,
                                                      "log_print_cb_set"));
  register_log_level_set(isolate, exports,
                         compatibility_new<String>(isolate, "log_level_set"));
  register_log_level_get(isolate, exports,
                         compatibility_new<String>(isolate, "log_level_get"));
  register_log_level_check(isolate, exports,
                           compatibility_new<String>(isolate,
                                                     "log_level_check"));
  register_log_color_disable_set(isolate, exports,
                                 compatibility_new<String>(isolate,
                                                           "log_color_disable"
                                                           "_set"));
  register_log_color_disable_get(isolate, exports,
                                 compatibility_new<String>(isolate,
                                                           "log_color_disable"
                                                           "_get"));
  register_log_file_disable_set(isolate, exports,
                                compatibility_new<String>(isolate,
                                                          "log_file_disable"
                                                          "_set"));
  register_log_file_disable_get(isolate, exports,
                                compatibility_new<String>(isolate,
                                                          "log_file_disable"
                                                          "_get"));
  register_log_abort_on_critical_set(isolate, exports,
                                     compatibility_new<String>(isolate,
                                                               "log_abort_on"
                                                               "_critical"
                                                               "_set"));
  register_log_abort_on_critical_get(isolate, exports,
                                     compatibility_new<String>(isolate,
                                                               "log_abort_on"
                                                               "_critical"
                                                               "_get"));
  register_log_function_disable_set(isolate, exports,
                                    compatibility_new<String>(isolate,
                                                              "log_function"
                                                              "_disable_set"));
  register_log_function_disable_get(isolate, exports,
                                    compatibility_new<String>(isolate,
                                                              "log_function"
                                                              "_disable_get"));
  register_log_abort_on_critical_set(isolate, exports,
                                     compatibility_new<String>(isolate,
                                                               "log_abort_on"
                                                               "_critical"
                                                               "_set"));
  register_log_abort_on_critical_get(isolate, exports,
                                     compatibility_new<String>(isolate,
                                                               "log_abort_on"
                                                               "_critical"
                                                               "_get"));
  register_log_abort_on_critical_level_set(isolate, exports,
                                           compatibility_new<String>(isolate,
                                                                     "log_abort"
                                                                     "_on"
                                                                     "_critical"
                                                                     "_level"
                                                                     "_set"));
  register_log_abort_on_critical_level_get(isolate, exports,
                                           compatibility_new<String>(isolate,
                                                                     "log_abort"
                                                                     "_on"
                                                                     "_critical"
                                                                     "_level"
                                                                     "_get"));
  register_log_domain_level_set(isolate, exports,
                                compatibility_new<String>(isolate,
                                                          "log_domain_level"
                                                          "_set"));
  register_log_domain_level_get(isolate, exports,
                                compatibility_new<String>(isolate,
                                                          "log_domain_level"
                                                          "_get"));
  register_log_state_start(isolate, exports,
                           compatibility_new<String>(isolate,
                                                     "LOG_STATE_START"));
  register_log_state_stop(isolate, exports,
                          compatibility_new<String>(isolate, "LOG_STATE_STOP"));
  register_log_timing(isolate, exports,
                      compatibility_new<String>(isolate, "log_timing"));

  eina_log_print_cb_set(eina_log_print_cb_js_test, NULL);
  eina_log_level_set(EINA_LOG_LEVEL_DBG);
  eina_log_abort_on_critical_set(EINA_FALSE);
}

#ifndef HAVE_NODEJS
int main(int, char*[])
{
  efl::eina::eina_init eina_init;
  efl::eo::eo_init eo_init;

  efl::eina::js::compatibility_initialize();
  v8::Isolate* isolate = v8::Isolate::New();
  assert(isolate != 0);

  v8::Isolate::Scope isolate_scope(isolate);
  
  efl::eina::js::compatibility_handle_scope handle_scope(isolate);
  v8::Handle<v8::Context> context
    = efl::eina::js::compatibility_new<v8::Context>
    (isolate, nullptr
     , efl::eina::js::compatibility_new<v8::ObjectTemplate>(isolate));
  if (context.IsEmpty()) {
    fprintf(stderr, "Error creating context\n");
    return 1;
  }
  context->Enter();
  {
    std::vector<char> script;
    {
      std::ifstream script_file(TESTS_SRC_DIR "/eina_js_suite.js");
      script_file.seekg(0, std::ios::end);
      std::size_t script_size = script_file.tellg();
      script_file.seekg(0, std::ios::beg);
      script.resize(script_size+1);
      script_file.rdbuf()->sgetn(&script[0], script_size);
      auto line_break = std::find(script.begin(), script.end(), '\n');
      assert(line_break != script.end());
      ++line_break;
      std::fill(script.begin(), line_break, ' ');

      std::cerr << "program:" << std::endl;
      std::copy(script.begin(), script.end(), std::ostream_iterator<char>(std::cerr));
      std::cerr << "end of program" << std::endl;
    }

    
    // Enter the execution environment before evaluating any code.
    v8::Context::Scope context_scope(context);
    v8::Local<v8::String> name(efl::eina::js::compatibility_new<v8::String>
                               (nullptr, TESTS_SRC_DIR "/eina_js_suite.js"));
    v8::Local<v8::Object> global = context->Global();
    v8::Handle<v8::Object> console = efl::eina::js::compatibility_new<v8::Object>(isolate);
    global->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "console"), console);
    console->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "log")
                 , efl::eina::js::compatibility_new<v8::FunctionTemplate>(isolate, & ::Print)
                 ->GetFunction()); 

    std::cerr << __LINE__ << std::endl;
    v8::Handle<v8::Object> exports = efl::eina::js::compatibility_new<v8::Object>(isolate);
    global->Set(efl::eina::js::compatibility_new<v8::String>(isolate, "suite"), exports);

    test_setup(exports);
    std::cerr << __LINE__ << std::endl;

    
    efl::eina::js::compatibility_handle_scope handle_scope(v8::Isolate::GetCurrent());
    std::cerr << __LINE__ << std::endl;
    ExecuteString(v8::Isolate::GetCurrent(),
                  efl::eina::js::compatibility_new<v8::String>(v8::Isolate::GetCurrent(), &script[0]),
                  name);
  std::cerr << __LINE__ << std::endl;
  }
  context->Exit();
}

#else
#include EINA_STRINGIZE(NODE_INCLUDE_HEADER)

namespace {

void eina_js_module_init(v8::Handle<v8::Object> exports)
{
  fprintf(stderr, "teste\n"); fflush(stderr);
  try
    {
      
      eina_init();
      eo_init();

      test_setup(exports);
      
      std::cerr << "registered" << std::endl;
    }
  catch(...)
    {
      std::cerr << "Error" << std::endl;
      std::abort();
    }
}
  
}

NODE_MODULE(eina_js_suite_mod, ::eina_js_module_init)

#endif

File: src/tests/eina_js/eina_js_suite.js
#!/usr/bin/env node

var suite;
var assert;

if(typeof process !== 'undefined')
{
    console.log('running from nodejs');
    console.log('path', process.env.NODE_PATH);
    console.log("teste1");

    suite = require('eina_js_suite_mod');
    assert = require('assert');
    assert(suite != null);
}
else
{
    assert = function(test, message) { if (test !== true) throw message; };
    console.log('running from libv8')
}

// container tests

console.log("teste");
var l1 = suite.raw_list;
console.log ("l1 ", l1.toString());
assert (l1.length == 3);
var l2 = suite.raw_list;
console.log ("l2 ", l2.toString());
assert (l2.length == 3);
var c = l1.concat(l2);
console.log ("c ", c.toString());
assert (c.length == (l1.length + l2.length));
assert (c[0] == l1[0]);
assert (c[1] == l1[1]);
assert (c[2] == l1[2]);
assert (c[3] == l2[0]);
assert (c[4] == l2[1]);
assert (c[5] == l2[2]);
assert (c.indexOf(c[0]) == 0);
assert (c.indexOf(c[1]) == 1);
assert (c.indexOf(c[2]) == 2);
assert (c.indexOf(c[3]) == 0);
assert (c.indexOf(c[4]) == 1);
assert (c.indexOf(c[5]) == 2);
assert (c.lastIndexOf(c[0]) == 3);
assert (c.lastIndexOf(c[1]) == 4);
assert (c.lastIndexOf(c[2]) == 5);
assert (c.lastIndexOf(c[3]) == 3);
assert (c.lastIndexOf(c[4]) == 4);
assert (c.lastIndexOf(c[5]) == 5);
var s1 = l1.slice(1, 3);
console.log ("s1 ", s1.toString());
assert (s1.length == 2);
assert (s1[0] == l1[1]);
assert (s1[1] == l1[2]);
var s2 = c.slice(1, 3);
console.log ("s2 ", s2.toString());
assert (s2.length == 2);
assert (s2[0] == l1[1]);
assert (s2[1] == l1[2]);

console.log("Finished containers");

// error tests

var captured = false;
console.log("#1");
try {
console.log("#2");
  suite.clear_eina_error();
console.log("#3");
} catch(e) {
console.log("#4");
  console.log("Exception ", e.toString());
console.log("#5");
  captured = true;
}
console.log("#6");
assert(captured === false, 'error #1');
console.log("#7");

captured = false;
try {
console.log("#1");
  suite.set_eina_error();
console.log("#2");
} catch(e) {
console.log("#3");
  console.log("Exception ", e.toString());
console.log("#4");
  assert(e.code === 'Eina_Error', 'error #2');
console.log("#5");
  assert(e.value === 'foobar', 'error #3');
console.log("#6");
  captured = true;
console.log("#7");
}
console.log("#8");
assert(captured === true, 'error #4');

console.log("Finished errors");

// accessor tests

assert(suite.acc.get(0) === 42, 'accessor #1');
assert(suite.acc.get(1) === 24, 'accessor #2');

console.log("Finished accessors");

// iterator tests

assert(suite.it.next().value === 42, 'iterator #1');
assert(suite.it.next().value === 24, 'iterator #2');

console.log("Finished iterators");

// value tests

console.log("x");

var my_value = new suite.value(1);
console.log("x");
var wrapped = my_value.get();
console.log("x");
assert(typeof(wrapped) === 'number', 'value #1');
console.log("x");
assert(wrapped === 1, 'value #2');
console.log("x");

my_value.set(2);
console.log("x");
assert(wrapped === 1, 'value #3');
console.log("x");
wrapped = my_value.get();
console.log("x");
assert(typeof(wrapped) === 'number', 'value #4');
console.log("x");
assert(wrapped === 2, 'value #5');
console.log("x");

my_value.set(true);
console.log("x");
assert(wrapped === 2, 'value #6');
console.log("x");
wrapped = my_value.get();
console.log("x");
// boolean is represented as integer in the efl::eina::value layer
console.log("x");
assert(typeof(wrapped) === 'number', 'value #7');
console.log("x");
assert(wrapped === 1, 'value #8');
console.log("x");

console.log("x");
var captured = false;
console.log("x");
try {
console.log("x");
  my_value.set({type: 'complex object'});
console.log("x");
} catch(e) {
console.log("x");
  assert(e.code === 'std::bad_cast', 'value #9');
console.log("x");
  captured = true;
console.log("x");
}
console.log("x");
assert(captured === true, 'value #10');
console.log("x");

console.log("x");
captured = false;
console.log("x");
try {
console.log("x");
    my_value = new suite.value({type: 'complex object'});
console.log("x");
} catch(e) {
  console.log("e.code ", e.code, ' ', typeof e);
  assert(e.code === 'std::bad_cast', 'value #11');
  console.log("x");
  captured = true;
  console.log("x");
}
assert(captured === true, 'value #12');

// log

function f1(){ suite.log_print(suite.LOG_DOMAIN_GLOBAL, suite.LOG_LEVEL_DBG, 'I changed again'); }; f1();
suite.log_print(suite.LOG_DOMAIN_GLOBAL, suite.LOG_LEVEL_CRITICAL, 'Cool to Hate');
function f2(){ suite.log_print(suite.LOG_DOMAIN_GLOBAL, suite.LOG_LEVEL_WARN, 'One Fine Day'); }; f2();
function f3(){ suite.log_print(suite.mydomain, suite.LOG_LEVEL_INFO, 'Never Gonna Find Me'); }; f3();

mydomain2 = suite.log_domain_register('mydomain2', '');
suite.log_domain_registered_level_set(mydomain2, suite.LOG_LEVEL_DBG);
assert(suite.log_domain_registered_level_get(mydomain2) === suite.LOG_LEVEL_DBG, '#1');

suite.log_print(mydomain2, suite.LOG_LEVEL_ERR, "The Kids Aren't Alright");

suite.log_domain_unregister(mydomain2);
mydomain2 = undefined;

suite.log_color_disable_set(true);
assert(suite.log_color_disable_get() === true, '#2');
suite.log_color_disable_set(false);
assert(suite.log_color_disable_get() === false, '#3');

suite.log_file_disable_set(true);
assert(suite.log_file_disable_get() === true, '#4');
suite.log_file_disable_set(false);
assert(suite.log_file_disable_get() === false, '#5');

suite.log_function_disable_set(true);
assert(suite.log_function_disable_get() === true, '#6');
suite.log_function_disable_set(false);
assert(suite.log_function_disable_get() === false, '#7');

suite.log_abort_on_critical_set(true);
assert(suite.log_abort_on_critical_get() === true, '#8');
suite.log_abort_on_critical_set(false);
assert(suite.log_abort_on_critical_get() === false, '#9');

var entered = false;
suite.log_print_cb_set(function(domain, color, level, file, func, line, msg) {
  assert(domain === 'mydomain', '#10');
  assert(color === '', '#11');
  assert(level === suite.LOG_LEVEL_WARN, '#12');
  assert(file === 'eina_js_suite.js', '#13');
  assert(func === 'f4', '#14');
  assert(line === 236, '#15');
  assert(msg === 'What Happened To You', '#16');
  entered = true;
});

function f4(){ suite.log_print(suite.mydomain, suite.LOG_LEVEL_WARN, 'What Happened To You'); }; f4();
assert(entered === true, '#17');

assert(suite.log_level_get() === suite.LOG_LEVEL_DBG, '#18');
assert(suite.log_level_check(suite.LOG_LEVEL_INFO) === true, '#19');
suite.log_level_set(suite.LOG_LEVEL_CRITICAL);
assert(suite.log_level_get() === suite.LOG_LEVEL_CRITICAL, '#20');
assert(suite.log_level_check(suite.LOG_LEVEL_INFO) === false, '#21');

assert(suite.log_abort_on_critical_get() === false, '#22');
suite.log_abort_on_critical_set(true);
assert(suite.log_abort_on_critical_get() === true, '#23');

suite.log_abort_on_critical_level_set(suite.LOG_LEVEL_CRITICAL);
assert(suite.log_abort_on_critical_level_get() == suite.LOG_LEVEL_CRITICAL, '#24');
suite.log_abort_on_critical_level_set(suite.LOG_LEVEL_ERR);
assert(suite.log_abort_on_critical_level_get() == suite.LOG_LEVEL_ERR, '#25');

suite.log_domain_level_set('mydomain', suite.LOG_LEVEL_WARN);
assert(suite.log_domain_level_get('mydomain') === suite.LOG_LEVEL_WARN, '#26');
suite.log_domain_level_set('mydomain', suite.LOG_LEVEL_INFO);
assert(suite.log_domain_level_get('mydomain') === suite.LOG_LEVEL_INFO, '#27');

assert(typeof(suite.LOG_STATE_START) === 'number', '#28');
assert(typeof(suite.LOG_STATE_STOP) === 'number', '#29');
assert(typeof(suite.log_timing) === 'function', '#30');

// finished tests

console.log ("Test execution with success");

